### ISSUE TITLE
[Bug] Control Key gets stuck.

### ISSUE THREAD

**Issue by jsmunroe** [cmderdev/cmder#2832]
### Version Information

```markdown
Cmder version: 221218
Operating system: Windows 10 x64
```

### Cmder Edition

Cmder Full (with Git)

### Description of the issue

The control key appears to get stuck as pressed. At some point while using Cmder, when I backspace it erases whole words and when I hit up or down to scroll through command history, it changes the existing line to "[A" for up and "[B" for down. Left and right are "[D" and "[C" respectively. I have to restart it when this happens. 

It has been happening rather frequently ever since the last update.

_After hitting Up key_
![image](https://user-images.githubusercontent.com/3356351/224161512-83c3e53a-a0e1-4bc4-981d-0d9107b8ce57.png)

### How to reproduce

It happens after a random amount of time. I do not know what action triggers it.

### Additional context

Full disclosure, I am not certain that this isn't ConEmu, Clink, or other third-party tools that Cmder uses. I'm not sure how I would be certain of that or why that is relevant. If someone had a problem with one of my projects that was a bug with something it was using, that is still a bug in my project. I would want it fixed even if it wasn't directly a problem with my project. But reports help you out too. 

I did search for existing problems within the ConEmu space and didn't find anything. 

### Checklist

- [X] I have read the documentation.
- [X] I have searched for similar issues and found none that describe my issue.
- [X] I have reproduced the issue on the latest version of Cmder.
- [X] I am certain my issues are not related to ConEmu, Clink, or other third-party tools that Cmder uses.

**Comment by DRSDavidSoft** [comment]
Hi there, thank you for reporting this bug. Can you please use the unmodified ConEmu for some amount of time to see if this issue also happens with it too? Please use Cmd.exe shell instead of clink if you can, to see if this issue is related to ConEmu or not.

Also, do you use any kind of software that might modify the keyboard hotkeys in any way, such as AutoHotkey? Have you checked if the Ctrl key of your keyboard is not actually registered as being pressed? Please test if this issue only happens in Cmder or not. Test your keyboard with utilities or online websites to see if the Ctrl is correctly being registered.

**Comment by chrisant996** [comment]
It sounds like the ConEmu terminal mode may be getting stuck in "XTerm" mode.

In the XTerm mode, arrow keys and other special keys get inserted in exactly the way you describe.

ConEmu tries to automatically figure out when to use Windows vs XTerm mode.
See ConEmu docs [Terminal input and output modes](https://conemu.github.io/en/TerminalModes.html) for details.

To make ConEmu show the current terminal mode:
1. Open the ConEmu Settings dialog.
2. In the tree view on the left, choose "Status bar" under the "Features" heading.
3. Make sure the "Show status bar" checkbox at the top is checked.
4. In the "Available columns" list, find and double-click on "Terminal modes" to add it to the status bar.
5. Click "Save Settings".

Now the status bar will show "WK" for the normal Windows mode, and "XK" for the XTerm mode.

**Comment by jsmunroe** [comment]
It shows Wk in the status bar when the problem exists. Here are the status bars for two windows, the top has the problem, the bottom is a brand new window that does not have the problem.

![image](https://github.com/cmderdev/cmder/assets/3356351/20c5e52b-6d23-47f5-a765-8a50d6e4580c)

**Comment by chrisant996** [comment]
@jsmunroe In any case, this seems like a [ConEmu](https://github.com/maximus5/conemu) issue, not a Cmder issue.

**Comment by chrisant996** [comment]
@jsmunroe Clink v1.5.1 was published a few minutes and clears the `ENABLE_VIRTUAL_TERMINAL_INPUT` mode even more aggressively than before.  Some console programs set it but neglect to clear it, and that's what causes the problem you reported.

Some console mode programs continue to run and use their parent's terminal input/output handles even after the parent (e.g. cmd.exe) regains control and shows an input prompt.  That's especially problematic, because they may set `ENABLE_VIRTUAL_TERMINAL_INPUT` again even after Clink has cleared it multiple times already (especially if a console mode program gets aborted e.g. via <kbd>Ctrl</kbd>-<kbd>C</kbd>, which can interfere with their ability to restore the console mode if they don't have proper error handling code).  Clink tries its best to compensate for those other buggy programs/scripts, but there is only so much that Clink can do to defend against them.

Ultimately, you need to track down which programs and/or scripts are triggering the problem, and then get them fixed.

**Comment by jsmunroe** [comment]
I guess it would have to be git since that's what I use most of the time.

**Comment by chrisant996** [comment]
> I guess it would have to be git since that's what I use most of the time.

Seems unlikely.  I use git all the time and I've only run into this 3 or 4 times in 2.5 years, and it was definitely with other programs _not_ git.

However, it could potentially be an external tool that's been configured to be spawned by git (e.g. a diff tool or merge tool or etc).

**Comment by chrisant996** [comment]
... or it's also possible that you're hitting something different that might be purely a ConEmu issue.  I don't use ConEmu.

By the way, I'm pretty sure the issue has nothing to do with the <kbd>Ctrl</kbd> key getting stuck.  It's the `ENABLE_VIRTUAL_TERMINAL_INPUT` console mode, which is something that can only be changed programmatically.  The VT100 escape codes for up, down, right, left are `ESC [ A`, `ESC [ B`, `ESC [ C`, and `ESC [ D`, respectively.

**Comment by daxgames** [comment]
That was a long standing bug in comemu that was fixed some time ago and I have not seen it since.

When it was an issue I saw it mostly in vim in a cmd.exe terminal.

**Comment by frankpfan** [comment]
I'm having the same problem on my Windows10 cmder full with ConEmu.

I'm used to `cd ..` via `Ctrl + Alt + U`. It has always worked fine, but not today.

I have another two problems meanwhile.

1. Cmder doesn't show the current git branch in the prompt. When I'm in a git repo, it fails to recognise and doesn't show `(main -> origin)`. However, git works fine.
2. Aliases aren't coloured in yellow. I use monokai color scheme, which shows alias names in a different colour. But now it's wrong. However, all alias commands are right.

So all of these seem to be bugs in ConEmu. They came to me at the same time. But there were no problems when I'd been using cmder just a short time ago. I suppose they're related?

**Additional Information**: When I open cmder, a line of message appear and immediately vanish, showing that it can't find clink_x64.exe. I only have clink_x86.exe and clink_arm64.exe in the right place. Would it be one cause?

*(Excuse my terrible English because I'm a Chinese student :)*

**Comment by DRSDavidSoft** [comment]
I have been trying several scenarios for these past couple of months and can confirm this is not related to the passage of an arbitrary amount of time; additionally it can also happen outside of Cmder and ConEmu. One way I have found to trigger it is to ssh into another host, run certain programs, and then hit <kbd>Ctrl</kbd> <kbd>Break</kbd> to kill the ssh connection. This will result in the same exact issue.

Currently I don't believe this issue is related to Cmder, it might be relevant to something else. Since it also happens with `cmd.exe` and `conhost.exe` (no Clink/ConEmu), I'm investigating to see exactly what is being changed in the terminal that causes this.

**Comment by chrisant996** [comment]
@DRSDavidSoft it's caused by what I said in [this comment](https://github.com/cmderdev/cmder/issues/2832#issuecomment-1622035180).

Some program is changing the console input mode, and not restoring the original console mode before exiting.

**Comment by DRSDavidSoft** [comment]
@chrisant996 Yeah I assumed that echoing something to the terminal might trigger this while the `ENABLE_VIRTUAL_TERMINAL_INPUT` flag is set.

Am I correct to assume according to your comment, just by setting the flag this issue can be repro'ed?

**Comment by chrisant996** [comment]
You're seeing echo'd _input_, not output. 

It's more complicated than just the flag being set.  But the flag is what allows the problem to occur. 

And the flag is present because a program explicitly forced it on because the program wants to receive emulated VT input, but the program neglected to restore the original flags before exiting.  So then other things can accidentally receive emulated VT input when they aren't expecting it. 

ConEmu has a built-in option for showing terminal modes in its status bar.  You can use ConEmu to watch for what program(s) sets the flag but exits without clearing it.

**Comment by KingLurchi** [comment]
For me this seems to happen only when I use certain git shortcuts, but not for all and never if I use the full command.

```
Command: git status --short --branch
Shortcut: git st

Command: git pull
Shortcut: git pl

Command: git fetch --all --prune
Shortcut: git fap
```

Once you enter any other git shortcut that does not trigger this issue, the problem goes away until you use one of the buggy ones again.
